AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java

```diff
@@ -33,18 +35,33 @@ public DismissType getDismissType() {
         return mDt;
     }
 
+
     /**
      * Return MULTI_CARD when no other action is needed, e.g. for multi card action
      * Return NO_REVIEW when we just need to reset the collection
+     * Return NOT_UNDONE if the undo was already started and so is not executed here
      * Returned positive integers are card id. Those ids is the card that was discarded and that may be sent back to the reviewer.*/
-    public abstract long undo(Collection col);
+    public long undo(Collection col){
+        synchronized (mStarted) {
+            if (mStarted) {
+                return NOT_UNDONE;
+            }
+            mStarted = true;
+        }
+        return actualUndo(col);
+    }
+    /**Actually undo. It assumes the action was not already undone. Return same as undo, but can't return NOT_UNDONE */
+    protected abstract long actualUndo(Collection col);
```

Comment:
(Optional) I think this feels slightly more descriptive. 
```suggestion
    protected abstract long performUndoOperation(Collection col);
```

Reply:
I'll change it if the boolean remain. If we only cancel head of the queue, then I won't need to add an extra member